spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
+        \union
+    {NoEvent}      
+
+(* Produce a request event for the top element of the requestStack *)
+RequestHeaderForTopRequest(stack) ==
+    IF stack = <<>>
+    THEN NoEvent
+    ELSE  LET top == Head(stack)
+              heightToRequest ==
+                IF top.isLeft
+                THEN top.endHeight      \* the pivot is on the right
+                ELSE top.startHeight    \* the pivot is on the left
           IN
-          blockchain = <<genesis>>
+          [type |-> "requestHeader", height |-> heightToRequest]
+
+(* When starting the light client *)
+OnStart ==
+    /\ state = "init"
+    /\ inEvent.type = "start"
+    /\ state' = "working"
+        (* the block at trusted height is obtained by the user *)
+        \* TODO: the English spec does not make this explicit
+    /\ storedHeaders' = { << blockchain[TRUSTED_HEIGHT],
+                             DOMAIN blockchain[TRUSTED_HEIGHT].VP >> }
+        (* The only request on the stack ("left", h1, h2) *)
+    /\ LET initStack == << [isLeft |-> TRUE,
+                           startHeight |-> TRUSTED_HEIGHT,
+                           endHeight |-> inEvent.heightToVerify] >>
+       IN           
+        /\ requestStack' = initStack
+        /\ outEvent' = RequestHeaderForTopRequest(initStack)
+
+(* Check whether we can trust the signedHdr based on trustedHdr
+   following the trusting period method.
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
+    IF minTrustedHeight > heightToTrust \* outside of the trusting period
+        \/ ~(signedHdr[2] \subseteq (DOMAIN signedHdr[1].VP)) \* signed by other nodes
+    THEN FALSE
+    ELSE
+      LET TP == BC!PowerOfSet(trustedHdr.NextVP, DOMAIN trustedHdr.NextVP)
+          SP == BC!PowerOfSet(trustedHdr.NextVP,
+            signedHdr[2] \intersect DOMAIN trustedHdr.NextVP)
+      IN
+      IF heightToVerify = heightToTrust + 1
+          \* the special case of adjacent heights: check 2/3s
+      THEN (3 * SP > 2 * TP)
+      ELSE \* the general case: check 1/3 between the headers and make sure there are 2/3 votes  
+        LET TPV == BC!PowerOfSet(signedHdr[1].VP, DOMAIN signedHdr[1].VP)
+            SPV == BC!PowerOfSet(signedHdr[1].VP, signedHdr[2])
+        IN
+        (3 * SP > TP) /\ (3 * SPV > 2 * TPV)
+
+(* Make one step of bisection, roughly one stack frame of Bisection in the English spec *)
+OneStepOfBisection(storedHdrs) ==
+    LET topReq == Head(requestStack)
+        lh == topReq.startHeight
+        rh == topReq.endHeight
+        lhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = lh
```

Comment:
Would be probably better using records instead of tuples if possible for readability purpose.

Reply:
done